[Liew Xin Yi] iP

xyliew25 opened the pull request:
### DukePro 
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/xyliew25/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the main method:

````java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
````

Commit:
Add Gradle support

xyliew25 committed:
complete Level 1

xyliew25 committed:
complete Level 2

xyliew25 committed:
fix bye response

xyliew25 committed:
add Task.java

xyliew25 committed:
complete Level 3

xyliew25 committed:
remove comments

xyliew25 committed:
add Todo.java

xyliew25 committed:
add Deadline.java

xyliew25 committed:
add Event.java

xyliew25 committed:
complete Level 4

xyliew25 committed:
fix task command response singular plural typo

xyliew25 committed:
close scanner

xyliew25 committed:
complete Level 4

xyliew25 committed:
add DukeException.java

xyliew25 committed:
add EmptyDescriptionException.java

xyliew25 committed:
add UnknownCommandException.java

xyliew25 committed:
add MissingArgumentException.java

xyliew25 committed:
add InvalidArgumentException.java

xyliew25 committed:
complete Level 5

xyliew25 committed:
update test cases

xyliew25 committed:
substitute array with arraylist

xyliew25 committed:
complete Level 6

xyliew25 committed:
update test cases

xyliew25 committed:
add toStringData method to Task and its subclasses

xyliew25 committed:
ignore data folder

xyliew25 committed:
complete Level 7

xyliew25 committed:
add InvalidDateFormatException.java

xyliew25 committed:
complete Level 8

xyliew25 committed:
update test cases

xyliew25 committed:
Merge branch 'branch-Level-7'

xyliew25 committed:
resolve merge conflict

xyliew25 committed:
add Command classes

xyliew25 committed:
add file exception classes

xyliew25 committed:
add more OOP classes

xyliew25 committed:
ignore data folder in text-ui-test folder

xyliew25 committed:
complete A MoreOOP

xyliew25 committed:
complete A Packages

xyliew25 committed:
update runtest.bat

xyliew25 committed:
complete A JUnit

xyliew25 committed:
Complete A-CodingStandard

xyliew25 committed:
complete A JavaDoc

xyliew25 committed:
Remove unnecessary this keyword

xyliew25 committed:
Complete Level 9

xyliew25 committed:
Update test cases

xyliew25 committed:
Merge branch 'branch-A-JavaDoc'

xyliew25 committed:
Resolve merge conflict

xyliew25 committed:
Resolve merge conflict

xyliew25 committed:
Merge branch 'add-gradle-support' of https://github.com/xyliew25/ip into add-gradle-support

xyliew25 committed:
Complete A-Gradle

xyliew25 committed:
Merge branch 'branch-A-Gradle'

xyliew25 committed:
Complete A-CheckStyle

xyliew25 committed:
Merge branch 'branch-A-CheckStyle'

xyliew25 committed:
Add appropriate final keywords

xyliew25 committed:
Complete Level 10

xyliew25 committed:
Merge branch 'branch-Level-10'

wei2912 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,40 @@
+     * @throws DukeException If there is/are invalid argument(s) passed into this command,
+     *                       or changes cannot be saved to storage.
+     */
+    public abstract String executeAndGetResponse(TaskList tasks, Ui ui, Storage storage) throws DukeException;
```
This function seems to be essentially equivalent to `execute` other than the fact that it has a `String` return type. Was there any specific reason why you chose to have a separate abstract function, rather than to change the return type of `execute` and return empty/null strings if there is no response?

Looking at the classes that extend from `Command`, it seems that this has led to multiple [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) violations. I think reworking your codebase to have just the `execute` method would be beneficial for code quality.

xyliew25 replied:
This was a quick fix after adding the GUI as the response of each command is to be returned, unlike in the CLI, the response is just to be printed on the console. However, your comment gave me an idea to refactor my code. Something as follows:
````java
public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
    String response = executeAndGetResponse(tasks, ui, storage);
    ui.showResponse(response);
}
````

Also, your comment made me realized that I should have mentioned its purpose in the header comment.

Thanks!

wei2912 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,82 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        TaskList matchingTasks = new TaskList();
+        for (int i = 0; i < tasks.getSize(); i++) {
+            Task currTask = tasks.getTask(i);
```
You might prefer to use the enhanced `for` loop here, i.e. `for (Task currTask : tasks)`. (This also applies to the loops in other command classes.)

xyliew25 replied:
Yes, thank you for the suggestion!

wei2912 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,82 @@
+                matchingTasks.addTask(currTask);
+            }
+        }
+        String response;
+        if (matchingTasks.getSize() == 0) {
+            response = "There are no matching tasks.";
+        } else {
+            response = "Here are the matching tasks in your list:\n" + matchingTasks;
+        }
```
I think the ternary operator (`x ? y : z`) can be used here to make the code slightly neater.

xyliew25 replied:
Yes, indeed it will be much neater and I won't need to declare `response` without assigning it with a value. Thanks!

wei2912 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    @Override
+    public String toStringData() {
+        return "D | " + super.toStringData() + " | " + this.by;
```
Consider using `String.format` which might be easier to read compared to string concatenation.

mweeruien replied:
I agree with @wei2912. The usage of `String.format()` can also be considered in other areas of your code.

xyliew25 replied:
Hm I see. Thank you for the suggestion!

wei2912 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+ * Represents the Deadline task in the Duke program.
+ */
+public class Deadline extends Task {
+    protected LocalDate by;
```
While this isn't technically a coding standard violation (to my knowledge), it might be clearer to name this variable with a noun, e.g. `date`.

xyliew25 replied:
I reserved this suggestion as I see this variable as the argument `by` of a `Deadline` task instead of the preposition 'by'. However, you do have a point as it does not sound like a noun.

wei2912 reviewed (commented):
Great work so far! I really like how you've structured your codebase, as creating specific command classes and exceptions would be beneficial to future expansion.

I have pointed out a few areas where there is room for improvement, let me know if you have any queries!

mweeruien commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,82 @@
+        if (matchingTasks.getSize() == 0) {
+            response = "There are no matching tasks.";
+        } else {
+            response = "Here are the matching tasks in your list:\n" + matchingTasks;
```
Would it be better to leave a line break in the middle for greater readability?

xyliew25 replied:
Thank you for the suggestion!

mweeruien commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,153 @@
+            // If pass all checking, create and return find command obj
+            return new FindCommand(keyword);
+
+        case Parser.DEADLINE_COMMAND:
+            // Check for empty description
+            if (commandComponents.length == 1) {
+                throw new EmptyDescriptionException("Deadline");
+            }
+            arguments = fullCommand.split(" ", 2)[1];
+            argumentComponents = arguments.split(" /by ");
+            // Extract description argument
+            description = argumentComponents[0];
+            // Check for missing by argument
+            if (argumentComponents.length == 1) {
+                throw new MissingArgumentException("Deadline", "/by");
+            }
+            // Extract and parse by argument
+            try {
+                by = LocalDate.parse(arguments.split(" /by ")[1]);
+            } catch (DateTimeParseException err) {
+                throw new InvalidDateFormatException();
+            }
+            // If pass all checking, create and return deadline command obj
+            return new DeadlineCommand(description, by);
```
I really like the comments, they help in understanding the code! However, would it be better to leave several line breaks here as well to improve readability? The other `case` blocks can also be reviewed to see if the addition of line breaks would improve readability.

xyliew25 replied:
Glad the comments help! Yes, I shall add in the line breaks.

mweeruien reviewed (commented):
Your code adheres to the coding standards well, and I think the comments and Javadocs that complement your code were also well-written! Though there are several small changes that can be considered to increase the readability, it was overall a good job!

xyliew25 committed:
Refactor code after PR peer review

xyliew25 committed:
Add assertions

Assertions are not used.

Assertions should be used to verify assumptions about the program to
detect potential bugs and terminate the program as soon as possible
upon assertion failures.

Let's add assertions after checkings and object creations.

Checkings and object creations impose assumptions.

xyliew25 committed:
Merge pull request #1 from xyliew25/branch-A-Assertions

Add assertions

xyliew25 committed:
Refactor code

Some parts of the code lack quality.

Quality code improves readability, reduces errors and enhances
understanding.

Let's use tenary operator for one liner if-else statements, include
default branch in case statements and rename method name to reflect
method logic.

These methods improve code quality.

xyliew25 committed:
Merge pull request #2 from xyliew25/branch-A-CodeQuality

Refactor code

xyliew25 committed:
Set up Continuous Integration

xyliew25 committed:
Relocate and rename .yml file

xyliew25 committed:
Update gradle.yml

xyliew25 committed:
Fix response indentation

xyliew25 committed:
Add sort command

xyliew25 committed:
Merge pull request #3 from xyliew25/branch-C-Sort

Add sort command

xyliew25 committed:
Update GUI

xyliew25 committed:
Add screenshot

xyliew25 committed:
Reduce font size of dialog

xyliew25 committed:
Add user guide

xyliew25 committed:
Fix user guide table

xyliew25 committed:
Fix trailing whitespaces bug as spotted by @fans2619

xyliew25 committed:
Fix code quality issues as detected by @cs2103-bot

xyliew25 committed:
Update README.md

xyliew25 committed:
Set theme jekyll-theme-cayman

xyliew25 committed:
Remove FAQ

xyliew25 committed:
Merge branch 'master' of https://github.com/xyliew25/ip

xyliew25 committed:
Configurate README.md

xyliew25 committed:
Fix README.md

xyliew25 committed:
Test out configurations

xyliew25 committed:
Update text-ui-test

xyliew25 committed:
Change description of LoadingException, SavingException and StorageException

xyliew25 committed:
Remove indentation for lines and error messages

xyliew25 committed:
Update test cases

xyliew25 committed:
Fix greeting identation
